File: Stack/valid parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closedToOpen = { ")" : "(", "]":"[", "}":"{" }

        for c in s:
            if c in closedToOpen:
                # not empty and last one is the opening bracket
                if stack and stack[-1] == closedToOpen[c]:
                    stack.pop()
                else:
                    return False 
            else:
                stack.append(c)


        return True if not stack else False

File: Stack/valid_parentheses.cpp
#include <string>
#include <stack>
#include <unordered_map>

class Solution {
public:
    bool isValid(std::string s) {
        std::stack<char> open;
        std::unordered_map<char, char> parens = {
            {')', '('},
            {']', '['},
            {'}', '{'},
        };

        for (const& c : s)
        {
            if (parens.find(c) != parens.end())
            {
                if (open.empty()) return false;

                if (open.top() != parens[c]) return false;

                open.pop();
            }
            else
            {
                open.push(c);
            }
        }

        return open.empty();
    }
};